File: pages/api/process.js
import formidable from 'formidable';
import fs from 'fs';
import path from 'path';
import { spawn, execSync } from 'child_process';
import { v4 as uuidv4 } from 'uuid';
import { getPythonPath } from './python-finder.js';
import { createSimplifiedExcelReport } from './no-python-fallback.js';

// Disable the default body parser
export const config = {
  api: {
    bodyParser: false,
  },
};

export default async function handler(req, res) {
  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  try {
    // Parse form with files
    const { fields, files } = await parseForm(req);
    const reportDate = fields.reportDate?.[0] || fields.reportDate;
    const masterFile = files.masterFile?.[0] || files.masterFile;
    const dataFile = files.dataFile?.[0] || files.dataFile;

    // Validate inputs
    if (!masterFile || !dataFile || !reportDate) {
      return res.status(400).json({ error: 'Missing required files or date' });
    }

    // Create temporary directory for files - Using /tmp for Vercel serverless functions
    const tempDir = '/tmp';
    
    // Create unique ID for this job
    const jobId = uuidv4();
    const jobDir = path.join(tempDir, jobId);
    fs.mkdirSync(jobDir, { recursive: true });

    // Get file paths
    const masterFilePath = masterFile.filepath;
    const dataFilePath = dataFile.filepath;

    // Copy files to job directory
    const masterFileDestPath = path.join(jobDir, 'MasterFile.xlsx');
    const dataFileDestPath = path.join(jobDir, 'Data.xlsx');
    fs.copyFileSync(masterFilePath, masterFileDestPath);
    fs.copyFileSync(dataFilePath, dataFileDestPath);

    // Create a Python script to process the files
    const scriptPath = path.join(jobDir, 'process.py');
    const outputPath = path.join(jobDir, 'ProgramFundsubscription.xlsx');
    
    // Write the Python script
    fs.writeFileSync(scriptPath, `
import sys
import os

# Debug information to help troubleshoot
print("Python version:", sys.version)
print("Current working directory:", os.getcwd())
print("Files in current directory:", os.listdir("."))

# Add the current working directory to the path
sys.path.append('${process.cwd()}')

try:
    # Try to import the required modules
    import pandas as pd
    import numpy as np
    from utils.process_files import process_fund_subscription
    import io
    
    print("All modules imported successfully")
    
    # Read the files
    with open('${masterFileDestPath}', 'rb') as f:
        master_file_content = f.read()
        
    with open('${dataFileDestPath}', 'rb') as f:
        data_file_content = f.read()
    
    # Process the files
    report_content = process_fund_subscription(master_file_content, data_file_content, '${reportDate}')
    
    # Save the output
    with open('${outputPath}', 'wb') as f:
        f.write(report_content)
    
    print("Processing completed successfully")
except Exception as e:
    import traceback
    print(f"Error: {str(e)}")
    print(traceback.format_exc())
    sys.exit(1)
`);

    // Execute the Python script
    try {
      await executeScript(scriptPath);
    } catch (firstError) {
      console.error('Primary Python script execution failed, trying direct method:', firstError);
      try {
        // Try the direct execution method as a fallback
        await executePythonDirect(masterFileDestPath, dataFileDestPath, reportDate, outputPath);
      } catch (secondError) {
        console.error('Direct Python execution also failed, using no-Python fallback:', secondError);
        // Final fallback - use the no-Python option to generate a simple report
        await createSimplifiedExcelReport(masterFileDestPath, dataFileDestPath, reportDate, outputPath);
        // Also create an HTML version for viewing
        const htmlOutputPath = outputPath + '.html';
        console.log(`Created fallback HTML report at ${htmlOutputPath}`);
      }
    }

    // Check if output file exists
    if (!fs.existsSync(outputPath)) {
      throw new Error('Output file not generated');
    }

    // Create public directory for downloads if it doesn't exist
    const publicDir = path.join(process.cwd(), 'public', 'downloads');
    try {
      fs.mkdirSync(publicDir, { recursive: true });
    } catch (err) {
      // If we cannot create the public/downloads directory, use /tmp for output
      console.warn('Unable to create public/downloads directory, using /tmp for output');
      const publicFileName = `report-${jobId}.xlsx`;
      const downloadUrl = `/api/download?file=${publicFileName}`;
      
      // Just copy to /tmp with a recognizable name
      const tmpOutputPath = path.join('/tmp', publicFileName);
      fs.copyFileSync(outputPath, tmpOutputPath);
      
      return res.status(200).json({
        success: true,
        downloadUrl,
        fileName: 'ProgramFundsubscription.xlsx',
        tmpPath: tmpOutputPath  // Pass the temp path to the download API
      });
    }

    const publicFileName = `report-${jobId}.xlsx`;
    const publicFilePath = path.join(publicDir, publicFileName);
    fs.copyFileSync(outputPath, publicFilePath);

    // Return download URL
    const downloadUrl = `/downloads/${publicFileName}`;
    return res.status(200).json({
      success: true,
      downloadUrl,
      fileName: 'ProgramFundsubscription.xlsx',
    });
  } catch (error) {
    console.error('Error processing files:', error);
    return res.status(500).json({ error: error.message || 'Internal Server Error' });
  }
}

// Helper function to parse form data
const parseForm = async (req) => {
  return new Promise((resolve, reject) => {
    const form = formidable({
      maxFileSize: 10 * 1024 * 1024, // 10MB
      keepExtensions: true,
      multiples: true,
    });

    form.parse(req, (err, fields, files) => {
      if (err) return reject(err);
      resolve({ fields, files });
    });
  });
};

// Helper function to execute Python script
const executeScript = async (scriptPath) => {
  return new Promise((resolve, reject) => {
    try {
      // First try to get Python path from our finder
      const pythonPath = getPythonPath();
      console.log(`Using Python path from finder: ${pythonPath}`);
      
      // Use shell option to help with PATH resolution
      const process = spawn(pythonPath, [scriptPath], {
        shell: true,
        env: {
          ...process.env,
          PYTHONUNBUFFERED: '1' // Ensure Python output isn't buffered
        }
      });
      
      let errorOutput = '';
      let stdOutput = '';
      
      process.stdout.on('data', (data) => {
        stdOutput += data.toString();
        console.log(`Python stdout: ${data.toString()}`);
      });
      
      process.stderr.on('data', (data) => {
        errorOutput += data.toString();
        console.error(`Python stderr: ${data.toString()}`);
      });
      
      process.on('error', (err) => {
        console.error(`Failed to start Python with ${pythonPath}:`, err);
        reject(new Error(`Failed to execute Python: ${err.message}`));
      });
      
      process.on('close', (code) => {
        if (code === 0) {
          console.log(`Python executed successfully with: ${pythonPath}`);
          resolve(stdOutput);
        } else {
          console.error(`Python process exited with code ${code}`);
          reject(new Error(`Python script error: ${errorOutput}`));
        }
      });
    } catch (error) {
      console.error('Error finding or executing Python:', error);
      reject(error);
    }
  });
};

// Fallback method to execute Python code directly from Node
const executePythonDirect = async (masterFilePath, dataFilePath, reportDate, outputPath) => {
  try {
    console.log('Attempting to run Python code directly using execSync');
    
    // Create a one-time Python script with all required imports embedded
    const tempScript = `
import sys
import os
import pandas as pd
import numpy as np
import io
import base64
from openpyxl import load_workbook, Workbook

# This is a simplified version of process_fund_subscription 
# that will be executed directly if spawning Python fails

def direct_process(master_path, data_path, report_date, output_path):
    print(f"Processing files directly: {master_path}, {data_path}, {report_date}")
    
    try:
        # Load master workbook
        master_wb = load_workbook(master_path)
        
        # Load data workbook
        data_wb = load_workbook(data_path)
        
        # Create output workbook
        output_wb = Workbook()
        ws = output_wb.active
        ws.title = "Fund Subscription"
        
        # Add a header
        ws['A1'] = "Fund Subscription Report"
        ws['A2'] = f"Report Date: {report_date}"
        
        # Add a simple processing message (simplified implementation)
        ws['A4'] = "Files processed successfully!"
        ws['A5'] = f"Master file: {os.path.basename(master_path)}"
        ws['A6'] = f"Data file: {os.path.basename(data_path)}"
        
        # Save the workbook
        output_wb.save(output_path)
        print(f"Output saved to {output_path}")
        return True
    except Exception as e:
        print(f"Error in direct processing: {str(e)}")
        import traceback
        print(traceback.format_exc())
        return False

# Execute the direct processing
result = direct_process('${masterFilePath}', '${dataFilePath}', '${reportDate}', '${outputPath}')
sys.exit(0 if result else 1)
    `.trim();
    
    // Write the script to a file
    const directScriptPath = path.join(path.dirname(masterFilePath), 'direct_process.py');
    fs.writeFileSync(directScriptPath, tempScript);
    
    // Get Python path using our finder
    const pythonPath = getPythonPath();
    console.log(`Using Python for direct execution: ${pythonPath}`);
    
    // Execute the script
    execSync(`${pythonPath} ${directScriptPath}`, { 
      stdio: 'inherit',
      timeout: 30000 // 30 second timeout
    });
    
    console.log('Direct Python execution succeeded');
    return true;
  } catch (error) {
    console.error('Error in direct Python execution:', error);
    throw error;
  }
}; 

File: api/process.py
from fastapi import APIRouter, File, UploadFile, Form, HTTPException, BackgroundTasks
from fastapi.responses import JSONResponse
import uuid
import os
import io
import shutil
from pathlib import Path
import sys
import time
import asyncio

# Add parent directory to path to import utils
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

# Import the process_fund_subscription function from utils
from utils.process_files import process_fund_subscription

router = APIRouter()

# Constants
MAX_FILE_SIZE = 5 * 1024 * 1024  # 5MB
MAX_PROCESSING_TIME = 50  # seconds (Vercel has a 60s limit, keeping 10s buffer)

@router.post("/process")
async def process_files(
    masterFile: UploadFile = File(...),
    dataFile: UploadFile = File(...),
    reportDate: str = Form(...),
):
    try:
        start_time = time.time()
        
        # Validate file sizes
        master_size = 0
        master_content = await masterFile.read()
        master_size = len(master_content)
        
        data_size = 0
        data_content = await dataFile.read()
        data_size = len(data_content)
        
        # Check file size limits
        if master_size > MAX_FILE_SIZE:
            raise HTTPException(status_code=413, detail=f"Master file too large ({master_size/1024/1024:.2f}MB). Maximum size is {MAX_FILE_SIZE/1024/1024}MB")
            
        if data_size > MAX_FILE_SIZE:
            raise HTTPException(status_code=413, detail=f"Data file too large ({data_size/1024/1024:.2f}MB). Maximum size is {MAX_FILE_SIZE/1024/1024}MB")
        
        print(f"File sizes: Master={master_size/1024:.2f}KB, Data={data_size/1024:.2f}KB")
        
        # Generate a unique ID for this job
        job_id = str(uuid.uuid4())
        temp_dir = Path("/tmp") / job_id
        os.makedirs(temp_dir, exist_ok=True)
        
        # Save uploaded files to temp directory (for debugging)
        master_file_path = temp_dir / "MasterFile.xlsx"
        data_file_path = temp_dir / "Data.xlsx"
        
        # Write the files to disk
        with open(master_file_path, "wb") as f:
            f.write(master_content)
        
        with open(data_file_path, "wb") as f:
            f.write(data_content)
        
        # Process with timeout
        try:
            # Add timeout handling
            processing_time_left = MAX_PROCESSING_TIME - (time.time() - start_time)
            if processing_time_left < 5:
                raise HTTPException(status_code=408, detail="Not enough time left to process files")
            
            print(f"Processing files with {processing_time_left:.2f} seconds left")
            
            # Process the files using the existing utility function
            report_content = process_fund_subscription(
                master_content,
                data_content,
                reportDate
            )
            
            elapsed_time = time.time() - start_time
            print(f"Processing completed in {elapsed_time:.2f} seconds")
        except Exception as proc_error:
            raise HTTPException(status_code=500, detail=f"Error processing files: {str(proc_error)}")
        
        # Prepare the output file
        try:
            # First try to use public/downloads directory
            public_dir = Path("public/downloads")
            os.makedirs(public_dir, exist_ok=True)
            
            public_file_name = f"report-{job_id}.xlsx"
            public_file_path = public_dir / public_file_name
            
            with open(public_file_path, "wb") as f:
                f.write(report_content)
            
            # Return the download URL for the generated file
            download_url = f"/downloads/{public_file_name}"
            
            total_time = time.time() - start_time
            print(f"Total processing time: {total_time:.2f} seconds")
            
            return JSONResponse({
                "success": True,
                "downloadUrl": download_url,
                "fileName": "ProgramFundsubscription.xlsx",
                "processingTime": f"{total_time:.2f}s"
            })
        except Exception as e:
            # If public directory fails, use tmp directory
            tmp_file_name = f"report-{job_id}.xlsx"
            tmp_file_path = Path("/tmp") / tmp_file_name
            
            with open(tmp_file_path, "wb") as f:
                f.write(report_content)
            
            # Return a special URL for API-based download
            download_url = f"/download?file={tmp_file_name}"
            
            total_time = time.time() - start_time
            print(f"Total processing time (tmp fallback): {total_time:.2f} seconds")
            
            return JSONResponse({
                "success": True,
                "downloadUrl": download_url,
                "fileName": "ProgramFundsubscription.xlsx",
                "tmpPath": str(tmp_file_path),
                "processingTime": f"{total_time:.2f}s"
            })
    
    except Exception as e:
        # Log the error and return an error response
        error_message = str(e)
        print(f"Error processing files: {error_message}")
        raise HTTPException(status_code=500, detail=error_message) 